File: server/src/test/java/rikka/shizuku/server/ConnectionHistoryTest.java
package rikka.shizuku.server;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;
import org.robolectric.shadows.ShadowProcess;
import java.io.File;
import java.nio.file.Files;
import java.util.List;
import static org.junit.Assert.*;

@RunWith(RobolectricTestRunner.class)
@Config(sdk = 34, manifest = Config.NONE)
public class ConnectionHistoryTest {
    @Rule public TemporaryFolder temporary = new TemporaryFolder();
    @Test public void latestConnectionSurvivesRestartAndProfilesStaySeparate() {
        ShadowProcess.setUid(2000);
        File file = new File(temporary.getRoot(), "connections.json");
        var history = new ConnectionHistory(file);
        var owner = ApplicationDiscoveryTest.app("same.app", 10123);
        var profile = ApplicationDiscoveryTest.app("same.app", 1010123);
        assertEquals(0, history.get(owner));
        history.connected(owner, 1000);
        history.connected(profile, 2000);
        history.connected(owner, 3000);
        var reloaded = new ConnectionHistory(file);
        assertEquals(3000, reloaded.get(owner));
        assertEquals(2000, reloaded.get(profile));
    }
    @Test public void reinstallOrUidReuseDoesNotInheritConnectionTime() {
        ShadowProcess.setUid(2000);
        var history = new ConnectionHistory(new File(temporary.getRoot(), "connections.json"));
        var app = ApplicationDiscoveryTest.app("example", 10123);
        history.connected(app, 1000);
        app.firstInstallTime++;
        assertEquals(0, history.get(app));
        app.firstInstallTime--;
        app.applicationInfo.uid++;
        assertEquals(0, history.get(app));
    }
    @Test public void pruningOneSuccessfullyEnumeratedUserPreservesOtherProfiles() {
        ShadowProcess.setUid(2000);
        File file = new File(temporary.getRoot(), "connections.json");
        var history = new ConnectionHistory(file);
        var owner = ApplicationDiscoveryTest.app("owner", 10123);
        var profile = ApplicationDiscoveryTest.app("profile", 1010123);
        history.connected(owner, 1000);
        history.connected(profile, 2000);
        history.pruneUser(0, List.of(), 3000);
        var reloaded = new ConnectionHistory(file);
        assertEquals(0, reloaded.get(owner));
        assertEquals(2000, reloaded.get(profile));
    }
    @Test public void oldSnapshotCannotRemoveNewConnectionAndLateWritesCannotMoveTimeBackwards() {
        ShadowProcess.setUid(2000);
        var history = new ConnectionHistory(new File(temporary.getRoot(), "connections.json"));
        var app = ApplicationDiscoveryTest.app("example", 10123);
        history.connected(app, 3000);
        history.connected(app, 2000);
        history.pruneUser(0, List.of(), 2500);
        assertEquals(3000, history.get(app));
    }
    @Test public void brokenHistoryDoesNotAffectNewConnections() throws Exception {
        ShadowProcess.setUid(2000);
        File file = temporary.newFile("connections.json");
        Files.writeString(file.toPath(), "broken json");
        var history = new ConnectionHistory(file);
        var app = ApplicationDiscoveryTest.app("example", 10123);
        assertEquals(0, history.get(app));
        history.connected(app, 1234);
        assertEquals(1234, new ConnectionHistory(file).get(app));
    }
    @Test public void rootWritesRemainOwnedByShell() throws Exception {
        ShadowProcess.setUid(0);
        File file = new File(temporary.getRoot(), "connections.json");
        var history = new ConnectionHistory(file);
        try (var os = org.mockito.Mockito.mockStatic(android.system.Os.class)) {
            history.connected(ApplicationDiscoveryTest.app("example", 10123), 1234);
            os.verify(() -> android.system.Os.fchown(org.mockito.ArgumentMatchers.any(java.io.FileDescriptor.class),
                    org.mockito.ArgumentMatchers.eq(2000), org.mockito.ArgumentMatchers.eq(2000)));
            os.verify(() -> android.system.Os.fchmod(org.mockito.ArgumentMatchers.any(java.io.FileDescriptor.class), org.mockito.ArgumentMatchers.eq(0600)));
        }
        assertEquals(1234, new ConnectionHistory(file).get(ApplicationDiscoveryTest.app("example", 10123)));
    }

}

File: server/src/test/java/eu/darken/porter/privileged/ConnectionHistoryTest.kt
package eu.darken.porter.privileged

import android.system.Os
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.eq
import org.mockito.Mockito.mockStatic
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import org.robolectric.shadows.ShadowProcess
import java.io.File
import java.io.FileDescriptor
import java.nio.file.Files

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34], manifest = Config.NONE)
class ConnectionHistoryTest {

    @get:Rule
    val temporary = TemporaryFolder()

    @Test
    fun latestConnectionSurvivesRestartAndProfilesStaySeparate() {
        ShadowProcess.setUid(2000)
        val file = File(temporary.root, "connections.json")
        val history = ConnectionHistory(file)
        val owner = installedApp("same.app", 10123)
        val profile = installedApp("same.app", 1010123)
        assertEquals(0, history.get(owner))
        history.connected(owner, 1000)
        history.connected(profile, 2000)
        history.connected(owner, 3000)
        val reloaded = ConnectionHistory(file)
        assertEquals(3000, reloaded.get(owner))
        assertEquals(2000, reloaded.get(profile))
    }

    @Test
    fun reinstallOrUidReuseDoesNotInheritConnectionTime() {
        ShadowProcess.setUid(2000)
        val history = ConnectionHistory(File(temporary.root, "connections.json"))
        val app = installedApp("example", 10123)
        history.connected(app, 1000)
        app.firstInstallTime++
        assertEquals(0, history.get(app))
        app.firstInstallTime--
        app.applicationInfo!!.uid++
        assertEquals(0, history.get(app))
    }

    @Test
    fun pruningOneSuccessfullyEnumeratedUserPreservesOtherProfiles() {
        ShadowProcess.setUid(2000)
        val file = File(temporary.root, "connections.json")
        val history = ConnectionHistory(file)
        val owner = installedApp("owner", 10123)
        val profile = installedApp("profile", 1010123)
        history.connected(owner, 1000)
        history.connected(profile, 2000)
        history.pruneUser(0, listOf(), 3000)
        val reloaded = ConnectionHistory(file)
        assertEquals(0, reloaded.get(owner))
        assertEquals(2000, reloaded.get(profile))
    }

    @Test
    fun oldSnapshotCannotRemoveNewConnectionAndLateWritesCannotMoveTimeBackwards() {
        ShadowProcess.setUid(2000)
        val history = ConnectionHistory(File(temporary.root, "connections.json"))
        val app = installedApp("example", 10123)
        history.connected(app, 3000)
        history.connected(app, 2000)
        history.pruneUser(0, listOf(), 2500)
        assertEquals(3000, history.get(app))
    }

    @Test
    fun brokenHistoryDoesNotAffectNewConnections() {
        ShadowProcess.setUid(2000)
        val file = temporary.newFile("connections.json")
        Files.writeString(file.toPath(), "broken json")
        val history = ConnectionHistory(file)
        val app = installedApp("example", 10123)
        assertEquals(0, history.get(app))
        history.connected(app, 1234)
        assertEquals(1234, ConnectionHistory(file).get(app))
    }

    @Test
    fun theFirstBareArrayFormatStillLoads() {
        ShadowProcess.setUid(2000)
        val file = temporary.newFile("connections.json")
        Files.writeString(file.toPath(), """[{"key":"0:example","uid":10123,"installed":500,"connected":1234}]""")
        val app = installedApp("example", 10123)
        app.firstInstallTime = 500
        val history = ConnectionHistory(file)
        assertEquals(1234, history.get(app))
        history.connected(app, 2000)
        assertTrue(Files.readString(file.toPath()).startsWith("{\"version\":1,"))
        assertEquals(2000, ConnectionHistory(file).get(app))
    }

    @Test
    fun anUnsupportedVersionStartsEmptyWithoutBreakingNewConnections() {
        ShadowProcess.setUid(2000)
        val record = """[{"key":"0:example","uid":10123,"installed":500,"connected":1234}]"""
        for (version in listOf("2", "1.5", "4294967297", "\"1\"", "null")) {
            val file = File(temporary.root, "connections-$version.json")
            Files.writeString(file.toPath(), """{"version":$version,"connections":$record}""")
            val app = installedApp("example", 10123)
            app.firstInstallTime = 500
            val history = ConnectionHistory(file)
            assertEquals(version, 0, history.get(app))
            history.connected(app, 2000)
            assertEquals(version, 2000, ConnectionHistory(file).get(app))
        }
    }

    @Test
    fun rootWritesRemainOwnedByShell() {
        ShadowProcess.setUid(0)
        val file = File(temporary.root, "connections.json")
        val history = ConnectionHistory(file)
        mockStatic(Os::class.java).use { os ->
            history.connected(installedApp("example", 10123), 1234)
            os.verify { Os.fchown(any(FileDescriptor::class.java), eq(2000), eq(2000)) }
            // 0600 in the Java; Kotlin has no octal literal.
            os.verify { Os.fchmod(any(FileDescriptor::class.java), eq(0x180)) }
        }
        assertEquals(1234, ConnectionHistory(file).get(installedApp("example", 10123)))
    }
}
